bin/amphora/src/amphora/node/concurrent_repository.rs

```diff
@@ -0,0 +1,60 @@
+use std::io;
+use std::time::Duration;
+
+use anyhow::Result;
+use async_trait::async_trait;
+use bytes::Bytes;
+use futures::Stream;
+use repository::{Repository, StreamInfo};
+
+use super::stringlock::StringLock;
+
+pub struct ConcurrentRepository {
+    key_locks: StringLock,
+    repo: Box<dyn Repository + Send + Sync>,
+}
+
+impl ConcurrentRepository {
+    pub fn new(
+        repo: Box<dyn Repository + Send + Sync>,
+        lifetime: Duration,
+        max_memory: usize,
+    ) -> Self {
+        Self {
+            key_locks: StringLock::new(lifetime).with_cleanup_trigger(max_memory),
+            repo,
+        }
+    }
+}
+
+#[async_trait]
+impl Repository for ConcurrentRepository {
+    async fn save(
+        &self,
+        id: String,
+        size: u64,
+        stream: Box<dyn Stream<Item = Result<Bytes, io::Error>> + Send + Sync + Unpin + 'static>,
+    ) -> Result<()> {
+        let mtx = self.key_locks.get_lock(&id).await;
+        let _w_guard = mtx.write().await;
+        self.repo.save(id, size, stream).await
+    }
+
+    async fn write(&self, id: String, range: interface::Range, body: Bytes) -> Result<u64> {
+        let mtx = self.key_locks.get_lock(&id).await;
+        let _w_guard = mtx.write().await;
+        self.repo.write(id, range, body).await
```

Comment:
It's a little bit out of scope, but I'm curious to know how you were thinking of handling the case where there is multiple write requests for a single large file (e.g. 10GB). Correct me if I'm wrong, but the way I understand it the second call to write will wait for the mutex to be available (which may take a very long time). So, should the calling function implements some kind of timeout or should it be implemented within the StringLock itself? (Sorry if this is a dumb question hahaha)

Reply:
C'est une bonne question en fait. Mon avis c'est que le serveur devrait probablement pas avoir d'opinion sur les timeouts à ce niveau là (c'est _valide_ d'avoir deux clients qui attendent pour écrire un fichier). Je crois que c'est au client de gérer la situation s'il ne veut _pas_ attendre, mais dans la majorité des cas on veut attendre (je crois).
e.g. tu veux pas que le client te retourne un timeouterror sur un write parce que y'a du attendre mettons 700ms. J'aimerais pas mal mieux que le write prenne 800ms mais finisse en success.

Reply:
Bon point, je suis d'accord. Ça fait du sens!